planning/behavior_path_lane_change_module/src/scene.cpp

```diff
@@ -842,226 +844,238 @@ PathWithLaneId NormalLaneChange::getPrepareSegment(
+    route_handler, current_lanes.back(), *lane_change_parameters_, direction_);
+
+  const auto dist_ego_to_current_lanes_center =
+    lanelet::utils::getLateralDistanceToClosestLanelet(current_lanes, current_pose);
+
+  // ignore object that are ahead of terminal lane change start
+  utils::path_safety_checker::filterObjects(objects, [&](const PredictedObject & object) {
+    const auto obj_polygon = tier4_autoware_utils::toPolygon2d(object).outer();
+    // ignore object that are ahead of terminal lane change start
+    auto distance_to_terminal_from_object = std::numeric_limits<double>::max();
+    for (const auto & polygon_p : obj_polygon) {
+      const auto obj_p = tier4_autoware_utils::createPoint(polygon_p.x(), polygon_p.y(), 0.0);
+      Pose polygon_pose;
+      polygon_pose.position = obj_p;
+      polygon_pose.orientation = object.kinematics.initial_pose_with_covariance.pose.orientation;
+      const auto dist = utils::getDistanceToEndOfLane(polygon_pose, current_lanes);
+      distance_to_terminal_from_object = std::min(dist_ego_to_current_lanes_center, dist);
+    }
+
+    return (minimum_lane_change_length > distance_to_terminal_from_object);
+  });
 }
 
-LaneChangeTargetObjectIndices NormalLaneChange::filterObject(
+void NormalLaneChange::filterObjectsByLanelets(
   const PredictedObjects & objects, const lanelet::ConstLanelets & current_lanes,
-  const lanelet::ConstLanelets & target_lanes,
-  const lanelet::ConstLanelets & target_backward_lanes) const
+  const lanelet::ConstLanelets & target_lanes, std::vector<PredictedObject> & current_lane_objects,
+  std::vector<PredictedObject> & target_lane_objects,
+  std::vector<PredictedObject> & other_lane_objects) const
 {
-  const auto current_pose = getEgoPose();
+  const auto & current_pose = getEgoPose();
   const auto & route_handler = getRouteHandler();
   const auto & common_parameters = planner_data_->parameters;
-  const auto minimum_lane_change_length = calcMinimumLaneChangeLength(
-    route_handler, current_lanes.back(), *lane_change_parameters_, direction_);
+  const auto check_optional_polygon = [](const auto & object, const auto & polygon) {
+    return polygon && isPolygonOverlapLanelet(object, *polygon);
+  };
 
-  // Guard
-  if (objects.objects.empty()) {
-    return {};
-  }
+  // get backward lanes
+  const auto backward_length = lane_change_parameters_->backward_lane_length;
+  const auto target_backward_lanes =
+    utils::getPrecedingLanelets(*route_handler, target_lanes, current_pose, backward_length);
 
-  // Get path
-  const auto path =
-    route_handler->getCenterLinePath(current_lanes, 0.0, std::numeric_limits<double>::max());
-  const auto target_path =
-    route_handler->getCenterLinePath(target_lanes, 0.0, std::numeric_limits<double>::max());
+  {
+    lane_change_debug_.current_lanes = current_lanes;
+    lane_change_debug_.target_lanes = target_lanes;
+
+    // TODO(Azu) change the type to std::vector<lanelet::ConstLanelet>
```

Comment:
Still TODO?

Reply:
Yes, this is debug topic, I need to fix debug marker, which I didn't prefer to do it in this PR.